### PR TITLE
feat(*) improve Global/Remote registration process

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

* Refactor Multicluster config:

```yaml
mode: standalone # ENV: KUMA_MODE By default autoconfigured to standalone

# ControlPlane mode
multicluster:
  global:
    # The Zone status poller timeout
    pollTimeout: 500ms # ENV: KUMA_MULTICLUSTER_GLOBAL_POLL_TIMEOUT
    kds:
      # Port of GRPC server that Local Kuma CP connects to
      grpcPort: 5685 # ENV: KUMA_MULTICLUSTER_GLOBAL_KDS_GRPC_PORT
      # Interval for regenerating Snapshot
      refreshInterval: 1s # ENV: KUMA_MULTICLUSTER_GLOBAL_KDS_REFRESH_INTERVAL
      # TlsCertFile defines a path to a file with PEM-encoded TLS cert.
      tlsCertFile: # ENV: KUMA_MULTICLUSTER_GLOBAL_KDS_TLS_CERT_FILE
      # TlsKeyFile defines a path to a file with PEM-encoded TLS key.
      tlsKeyFile: # ENV: KUMA_MULTICLUSTER_GLOBAL_KDS_TLS_KEY_FILE
  remote: # Remote CP configuration
    zone: "" # ENV: KUMA_MULTICLUSTER_REMOTE_ZONE the name of the zone for this Remote CP
    # GlobalAddress URL of Global Kuma CP
    globalAddress: # ENV KUMA_MULTICLUSTER_REMOTE_GLOBAL_ADDRESS
    kds:
      # Interval for regenerating Snapshot
      refreshInterval: 1s # ENV: KUMA_MULTICLUSTER_REMOTE_KDS_REFRESH_INTERVAL
      # RootCAFile defines a path to a file with PEM-encoded Root CA. Client will verify server by using it.
      rootCaFile: # ENV: KUMA_MULTICLUSTER_REMOTE_KDS_ROOT_CA_FILE
```

* `kubectl install dns` defaults to `$HOME/.kube/config`
* `kubectl install dns --port` to set the port of the DNS server
* `kubectl install dns` - fixed to set the port for KubeDNS
* permissions to update/patch/delete resources in Remote CP
* configure the polling timeout and default to 500ms

Add a Zone in Universal:

```bash
echo "type: Zone
name: zone-1
ingress:
  address: 192.168.0.1:1000" | ./build/artifacts-darwin-amd64/kumactl/kumactl apply -f -
```

Get Zones resource:
```bash
$ curl localhost:5681/zones
{
 "total": 1,
 "items": [
  {
   "type": "Zone",
   "name": "zone-1",
   "creationTime": "2020-07-22T19:37:28.442793+03:00",
   "modificationTime": "2020-07-22T19:37:28.442793+03:00",
   "ingress": {
    "address": "192.168.0.1:1000"
   }
  }
 ],
 "next": null
}
```

Get status:
```bash
$ curl localhost:5681/status/zones
[
 {
  "name": "zone-1",
  "url": "192.168.0.1:1000",
  "active": false
 }
]
```

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/246)

### CODE REVIEW

**Comment by bloqhead** [comment]
@nickolaev I've created a [PR](https://github.com/kumahq/kuma-gui/pull/45) in the GUI repo for this. Currently, the Remote CPs page is simply pulling data from `:5681/status/zones` but once we have the ability to do something like `:5681/zones/zone-1`, I can have the tabbed view below the table list displaying more detailed information (similar to the Dataplanes view that fetches from both Dataplanes and Dataplane Insights):

File: pkg/config/app/kuma-cp/kuma-cp.defaults.yaml [hunk 1]
```diff
@@ -261,17 +254,27 @@ dnsServer:
...
-    rootCaFile: # ENV: KUMA_KDS_CLIENT_ROOT_CA_FILE
+
+# ControlPlane mode
+multicluster:
+  global:
+    # The Zone status poller timeout
+    pollTimeout: 500ms # ENV: KUMA_MULTICLUSTER_GLOBAL_POLL_TIMEOUT
+    kds:
+      # Port of GRPC server that Local Kuma CP connects to
+      grpcPort: 5685 # ENV: KUMA_MULTICLUSTER_GLOBAL_KDS_GRPC_PORT
```
Comment (jakubdyszkiewicz) [inline pkg/config/app/kuma-cp/kuma-cp.defaults.yaml:265]: this could be just `port`

File: pkg/config/multicluster/multicluster.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// Multicluster configuration
+type MulticlusterConfig struct {
+	Global *GlobalConfig `yaml:"global,omitempty"`
+	Remote *RemoteConfig `yaml:"remote,omitempty"`
+}
+
+func (m *MulticlusterConfig) Sanitize() {
+	if m.Global != nil {
```
Comment (jakubdyszkiewicz) [inline pkg/config/multicluster/multicluster.go:80]: we don't if this for other pointers in config. I don't think there is a situation where it is nil

File: pkg/config/multicluster/multicluster.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		m.Remote.Sanitize()
+	}
+}
+
+func (m *MulticlusterConfig) Validate() error {
+	var result error
+	if m.Global != nil {
+		err := m.Global.Validate()
+		if err != nil {
+			_ = multierror.Append(result, err)
```
Comment (jakubdyszkiewicz) [inline pkg/config/multicluster/multicluster.go:93]: I think this should be `result = multierror.Append(result, err)`

File: test/e2e/kuma_deploy_hybrid_test.go [hunk 4]
```diff
@@ -234,4 +243,27 @@ metadata:
 		Expect(err).ToNot(HaveOccurred())
 		Expect(stdout).To(ContainSubstring("HTTP/1.1 200 OK"))
 	})
+
+	It("should sync traffic permissions", func() {
```
Comment (jakubdyszkiewicz) [inline test/e2e/kuma_deploy_hybrid_test.go:247]: What is this test about?
Is it about syncing resources? In this case, I think it makes more sense to check what is the state of CPs.
Is it about testing traffic permissions? I think it makes more sense to check it in a test file dedicated to TrafficPermission

File: test/e2e/kuma_deploy_hybrid_test.go [hunk 5]
```diff
@@ -234,4 +243,27 @@ metadata:
...
+		err := global.GetKumactlOptions().KumactlDelete("traffic-permission", "traffic-permission-all")
+		Expect(err).ToNot(HaveOccurred())
```
Comment (jakubdyszkiewicz) [inline test/e2e/kuma_deploy_hybrid_test.go:249]: This should be cleaned up by previous test

File: test/e2e/kuma_deploy_hybrid_test.go [hunk 6]
```diff
@@ -234,4 +243,27 @@ metadata:
...
+
+		err = YamlUniversal(trafficPermissionToKuma2)(global)
```
Comment (jakubdyszkiewicz) [inline test/e2e/kuma_deploy_hybrid_test.go:251]: `trafficPermissionToKuma2` how about better name? `all-to-2-remote`

File: test/e2e/kuma_deploy_hybrid_test.go [hunk 7]
```diff
@@ -234,4 +243,27 @@ metadata:
...
+		Expect(err).ToNot(HaveOccurred())
+
+		// Remote 3
```
Comment (jakubdyszkiewicz) [inline test/e2e/kuma_deploy_hybrid_test.go:254]: Is there `Remote 3`? I thought there was only Remote 1 and Remote 2

File: pkg/config/app/kuma-cp/kuma-cp.defaults.yaml (hunk 1 above)
Comment (nickolaev) [reply pkg/config/app/kuma-cp/kuma-cp.defaults.yaml:265]: `grpcPort` is commonly used throughout the code, e.g. 'xds', 'sds', 'mads'

File: test/e2e/kuma_deploy_hybrid_test.go (hunk 4 above)
Comment (nickolaev) [reply test/e2e/kuma_deploy_hybrid_test.go:247]: The primary goal was to check resource synchronization. But we also check how traffic permissions behave cross cluster.

File: test/e2e/kuma_deploy_hybrid_test.go (hunk 7 above)
Comment (nickolaev) [reply test/e2e/kuma_deploy_hybrid_test.go:254]: In Hybrid - 3 and 4 are the Docker/Universal clusters

File: pkg/config/loader_test.go [hunk 8]
```diff
@@ -215,24 +216,23 @@ general:
 guiServer:
   port: 8888
   apiServerUrl: http://localhost:1234
-mode:
-  mode: remote
+mode: remote
+multicluster:
   global:
```
[the remark this answers is not in the document]
Comment (nickolaev) [reply pkg/config/loader_test.go:221]: Instead I call the Validation only for the relevant config - ie global.validat() when in Global mode and so on.

[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/config/loader_test.go:219]: Yes, but this is not the same. I'm talking about a scenario when the user sets both remote and global, but that hundred percent is not what he wants

[the remark this answers is not in the document]
Comment (nickolaev) [reply pkg/config/loader_test.go:219]: We do not get NIL pointers during YAML unmarshalling. It always creates the empty structs.

**Pull request merged**